File: go/hot_100/15_3sum.go
package main

import "sort"

func threeSum(nums []int) [][]int {
	hashmap := make(map[int][]int)
	var res [][]int
	sort.Ints(nums)
	for i, v := range nums {
		if _, ok := hashmap[v]; ok {
			hashmap[v] = append(hashmap[v], i)
		} else {
			hashmap[v] = []int{i}
		}
	}
	for i := 0; i < len(nums); {
		for j := i + 1; j < len(nums); {
			if arr, ok := hashmap[-nums[i]-nums[j]]; ok {
				for _, k := range arr {
					if k > i && k > j {
						res = append(res, []int{nums[i], nums[j], nums[k]})
						break
					}
				}
			}
			j++
			for j < len(nums) && nums[j] == nums[j-1] {
				j++
			}
		}
		i++
		for i < len(nums) && nums[i] == nums[i-1] {
			i++
		}
	}
	return res
}

// func main() {
// 	threeSum([]int{0, 0, 0, 0})
// }

File: python/hot_100/15_3sum.py
from typing import List
from collections import defaultdict
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums=sorted(nums)
        dd=defaultdict(list)
        for i,num in enumerate(nums):
            dd[num].append(i)
        res=[]
        for i in range(len(nums)):
            if i!=0 and nums[i]==nums[i-1]:
                continue
            for j in range(i+1,len(nums)):
                if j!=i+1 and nums[j]==nums[j-1]:
                    continue
                two=-(nums[i]+nums[j])
                if two in dd.keys():
                    for k in dd[two]:
                        if k>i and k>j:
                            res.append([nums[i],nums[j],nums[k]])
                            break
        return res

s=Solution()
print(s.threeSum([0,0,0,0]))